### PR TITLE
Add extent support. Refs #1227

### PR DESCRIPTION
**Pull request by tmcw**
This works on my example style.

* Adds a cache for the typed arrays that set extents
* Fixes drawing ratio of lines to match different extents

Main point of uncertainty:

* This adds support for **a single extent value per vector tile**. While
  multiple values are technically permitted there aren't any tiles that
  create them yet.

Could use a review /cc @ansis @jfirebaugh @kkaefer

### CODE REVIEW

**Comment by jfirebaugh** [comment]
For terminology, we should reserve the word "buffer" for the GL concept and call the backing arrays "arrays" or "buffer arrays". We probably don't do this elsewhere but we should.

I don't think it's worth caching the array creation -- it'll be heavily dominated by `gl.createBuffer` / `gl.bufferData`. If you want to cache something, cache the buffer.

**Comment by tmcw** [comment]
Okay - given that caching the buffer seems pretty tricky and this seems to perform pretty well, i'll remove the cache.

**Comment by tmcw** [comment]
Removed cache in 598891b

**Comment by kkaefer** [comment]
> This adds support for a single extent value per vector tile. While multiple values are technically permitted there aren't any tiles that create them yet.

Is there major effort associated with doing it per-layer? While there currently may not be vector tiles that specify multiple extents, compositing vector tiles may easily produce such tiles.

**Comment by tmcw** [comment]
@kkaefer mostly it's just extremely hard to follow how a "source layer" becomes what we actually paint and where i would attach this data to make that possible. If you can give a pointer or two in that direction it's probably not terribly hard to implement.

**Comment by tmcw** [comment]
@ansis @kkaefer Dove into multi-extent-per-tile in https://github.com/mapbox/mapbox-gl-js/pull/1236 and figured out that it entails a much larger refactor, since we'd have to split Tile into two classes, one for the tile id and the other for buffers & transforms.

I think we should land this PR now, refactor & document gl-js, and then implement support for multiple extents in a single tile. Changed this one to Refs #1227
